plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java

```diff
@@ -188,6 +192,50 @@ public FileHiveMetastore(NodeVersion nodeVersion, HdfsEnvironment hdfsEnvironmen
         }
     }
 
+    @Override
+    public synchronized MetastoreLock acquireTableExclusiveLock(
+            HiveIdentity identity,
+            AcidTransactionOwner transactionOwner,
+            String queryId,
+            String dbName,
+            String tableName)
+    {
+        SchemaTableName schemaTableName = new SchemaTableName(dbName, tableName);
+
+        ReentrantLock lock = tableLocks.computeIfAbsent(schemaTableName, ignored -> new ReentrantLock());
+        lock.lock();
+
+        tableLocks.put(schemaTableName, lock);
+        return new FileMetastoreLock(schemaTableName);
+    }
+
+    private class FileMetastoreLock
+            implements MetastoreLock
+    {
+        private final SchemaTableName tableName;
+        boolean released;
+
+        public FileMetastoreLock(SchemaTableName tableName)
+        {
+            this.tableName = requireNonNull(tableName, "tableName is null");
+        }
+
+        @Override
+        public void release()
+        {
+            if (released) {
+                throw new IllegalStateException("Table lock has already been released");
+            }
+            released = true;
+            ReentrantLock lock = tableLocks.get(tableName);
+            // Currently, metastore locks are always acquired and released in the same thread.
```

Comment:
I don't get this, if they are always acquired and released in the same thread what is the point of using locks instead of booleans?

Reply:
Multiple threads can request the lock but whichever thread acquires it, must be the one to release it because of how ReentrantLock works. If the lock was acquired for the entire query, instead of just the commit block, it'd be harder to ensure that.